### PR TITLE
Autoscrolling!

### PR DESCRIPTION
**Pull request by alexlitty**
Hiya @maximecb :)

It turns out the browser doesn't send pointer movement events when you're autoscrolling (modifying `scrollLeft` & `scrollTop`), so this turned out to be super tricky!

I've finally landed nicely with this: There is a new handler that processes any pointer movements, whether they come from browser events or manual autoscrolling adjustments. That handler queues up a call to `adjustView` which decides whether autoscrolling should happen and, if it should, re-calls the new pointer movement handler to loop it all.

The autoscrolling speed / "magnitude" is based on how far away your pointer is to the editor's edge, which I think gives it a nice smoothness. If you're beyond the edge (outside the tab / browser window) you'll scroll much faster, but that has a capped speed so it doesn't get out of control.

This works for both moving and selecting nodes, though autoscrolling while selecting nodes might need some more care. I notice it "snaps" a little bit if you're moving the pointer quickly in some situations (I think because the graph is resizing? Need to look into it more).

Thoughts?

### CODE REVIEW

File: public/editor.js [hunk 1]
```diff
@@ -44,11 +44,17 @@ export class Editor
...
+        // Positional information about the pointer
+        this.pointer = {
+            // Currently known location of the pointer. This is provided by
+            // browser events and manually updated as needed, while
+            // autoscrolling for example.
+            position: null,
+
+            // Location of the pointer when it was last pressed down. Cleared
+            // when the pointer is no longer pressed down.
+            downPosition: null
```
Comment (maximecb) [inline public/editor.js:56]: I would maybe rename these pos and downPos, otherwise it makes for very long variable names.

**Comment by maximecb** [comment]
I tested it and it seems to work well, though you said you notice odd behavior sometimes, it would be nice to fix that.

As a general comment I think the code needs more commenting. It's not necessarily obvious what each function does. The adjustView method is pretty complex too so probably needs some commenting inside. Not clear what the padding constant is about for instance.

Can you explain why you need the timeout mechanism and queueing? I wish the code was simpler.

**Comment by alexlitty** [comment]
We can do much simpler! I'll refactor this and get more comments down.

I do think we will still need something like the timeout mechanism to do any autoscrolling. The core of this feature, I'm seeing, is that we need to do adjust scrollLeft and scrollTop when the mouse is doing something at the edge of the editor, given only "absolute" pointer movement events (when user input causes the absolute mouse position to change).

So it's up to us to implement a mechanism that performs the autoscroll over time, starting with as little as a single pointer movement event. My proposal here is to have a function (`adjustView`) that is first triggered by a pointer movement towards the editor edge, which moves the editor view slowly. It keeps calling itself on a delay, as long as we think the pointer remains at the edge. The final effect is a bunch of slow view adjustments over time, an autoscroll.

As for the queueing (still thinking of a better name for this, it's only sort of queueing), the idea is to stop multiple adjustments from happening per frame / per pointer event, and to make the scrolling smooth.

- If you have five mouse movements in one event loop iteration that are all at the same edge of the editor, we don't want to make five simultaneous adjustments in that direction, or else your autoscrolling speed will speed up and slow down based on how many mouse events your browser decides to emit (causing jitter).

- If you have one mouse movement that triggers adjustView at 0ms, adjustView will try re-calling itself at 16ms. If you have another mouse movement at 5ms, we don't want to queue _another_ call to adjustView at 5+16ms. Otherwise the autoscrolling would be happening at twice the speed (and with an unpleasant jitter), just because you moved your mouse a little bit.

Lots of subtleties in this one! If you have a different vision here, I'm all for that too of course.

**Comment by maximecb** [comment]
I think you're right that some kind of timeout mechanism is needed for smooth scrolling. It makes sense.

In terms of commenting I'd add a comment above every new method to explain what they do, explaining the logic in adjustView, and maybe adjustView could be renamed to scrollView or moveView, seems more descriptive?

You also mentioned there might be some bug that may be snappy. I haven't experienced it, but if there's a bug, let's fix it.

**Pull request closed without merging**